### PR TITLE
Only point to specific question and back when editing user input question

### PR DESCRIPTION
**Pull request by benbowler**
You can now edit a single question from both the user-input and settings pages and return instantly to the previous page without having to go through the entire question flow.

## Summary

Addresses issue #2499

## Relevant technical choices

I had to implement this quite differently to the IB because #2041 used the query string, not URL hashes for its implementation. I achieved the same outcome. One larger change I had to make was moving the `submitChanges` logic from the `UserInputPreview` component to the parent `UserInputQuestionnaire` component so that a question can now trigger a submit. 

## Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

### CODE REVIEW

**Comment by benbowler** [comment]
As this is blocked by #2041 I've created this PR against that branch until it's complete.

File: assets/js/components/user-input/UserInputQuestionnaire.js
Comment (eugene-manuilov) [inline assets/js/components/user-input/UserInputQuestionnaire.js:55]: @benbowler let's move this hook to the `UserInputQuestionWrapper` component to escape a need to pass it via props. Just destruct the first value only in the wrapper component: `const [ single ] = useQueryArg( ... )`

File: assets/js/components/user-input/UserInputQuestionnaire.js
Comment (eugene-manuilov) [inline assets/js/components/user-input/UserInputQuestionnaire.js:92]: @benbowler this functionality doesn't seem to be needed in the questionary component. Let's add a new callback in the preview component that will set this and call the `goTo` function passed to the component via props.

Also, please, use the `useQueryArg` hook in the preview component to get the setter function instead of passing it via props.

File: assets/js/components/user-input/UserInputQuestionnaire.js
Comment (eugene-manuilov) [inline assets/js/components/user-input/UserInputQuestionnaire.js:252]: @benbowler the `goToPreview` property isn't needed for the `UserInputPreview` component. Let's get rid of it.

File: assets/js/components/user-input/UserInputQuestionnaire.js
Comment (eugene-manuilov) [inline assets/js/components/user-input/UserInputQuestionnaire.js:153]: @benbowler instead of passing it to every wrapper, let's leverage `next` and `nextLabel` props. It can be done like this:

```javascript
let nextCallback = next;
let nextLabel = undefined;

if ( single === 'user-input' ) {
    nextCallback = goToPreview;
    nextLabel = __( 'Update', 'google-site-kit' );
} else if ( single === 'settings' ) {
    nextCallback = submitChanges;
    nextLabel = __( 'Submit', 'google-site-kit' );
}

...

return (
    ...
    <UserInputQuestionWrapper
        next={ nextCallback }
        nextLabel={ nextLabel }
    />
    ...
);
```

> For the last question you can use `nextLabel={ nextLabel || __( 'Preview', 'google-site-kit' ) }` to make sure it works correctly.

In this case, we won't need additional logic in the wrapper component.

**Review by eugene-manuilov — CHANGES_REQUESTED** [review]
Thanks, @benbowler! It's a good start, but we need to adjust your logic a little bit. Please, check the comments below and let me know if you need additional details.

File: assets/js/components/user-input/UserInputPreview.js
Comment (benbowler) [reply assets/js/components/user-input/UserInputPreview.js:53]: Is this format of prop destructuring a project-wide standard?

Comment (eugene-manuilov) [reply assets/js/components/user-input/UserInputPreview.js:53]: @benbowler I don't think we have a project-wide standard for it but in general, it's much easier to read if props with many elements are destructed in the function body.

**Comment by benbowler** [comment]
Hey @eugene-manuilov, thanks for the comments, I've been through and updated the PR.

File: assets/js/components/user-input/UserInputPreview.js
Comment (eugene-manuilov) [inline assets/js/components/user-input/UserInputPreview.js:63]: @benbowler :man_facepalming: excuse me for the incorrect suggestion at the beginning, but we need to move error handling back to the `UserInputQuestionnaire` component and render it there instead of keeping it in the preview component. The idea here is that we need to display error no matter which screens we are on (a single question or the preview).

File: assets/js/components/user-input/UserInputPreview.js
Comment (eugene-manuilov) [inline assets/js/components/user-input/UserInputPreview.js:73]: The `single` value isn't needed here. We can simplify it to be:

```suggestion
	const [ , setSingle ] = useQueryArg( 'single', false );

	const editSingle = useCallback( ( questionNumber ) => {
		setSingle( 'user-input' );
		goTo( questionNumber );
	}, [] );
```

File: assets/js/components/user-input/UserInputQuestionWrapper.js
Comment (eugene-manuilov) [inline assets/js/components/user-input/UserInputQuestionWrapper.js:112]: Let's remove comments:

```suggestion
```

File: assets/js/components/user-input/UserInputQuestionWrapper.js
Comment (eugene-manuilov) [inline assets/js/components/user-input/UserInputQuestionWrapper.js:87]: Let's use the same approach as we use for the next callback in the `UserInputQuestionnaire` component:

```javascript
	// Update the callbacks and labels for the questions if the user is editing a *single question*.
	let backCallback = back;
	let nextCallback = next;
	let nextLabel;

	if ( single === 'user-input' ) {
		// When the user is editing a single question in the user-input screen send them back to the preview when they click Update.
		nextCallback = goToPreview;
		nextLabel = __( 'Update', 'google-site-kit' );
		backCallback = undefined;
	} else if ( single === 'settings' ) {
		// When the user is editing a single question from the settings screen, submit changes and send them back to the settings pages when they click Submit.
		nextCallback = submitChanges;
		nextLabel = __( 'Submit', 'google-site-kit' );
		backCallback = undefined;
	}
```

In this case we can escape changing the wrapper component at all.

File: assets/js/components/user-input/UserInputQuestionnaire.js
Comment (eugene-manuilov) [inline assets/js/components/user-input/UserInputQuestionnaire.js:247]: We need to display this progress bar not only when we are on the preview step. We need to display it when we are saving a single question as well. Let's move it above, something like this:

```javascript
if ( isSavingSettings || isNavigating ) {
	return (
		<ProgressBar />
	);
}

return (
	<Fragment>
		<ProgressBar
			height={ 0 }
			indeterminate={ false }
			progress={ ( activeSlugIndex + 1 ) / USER_INPUT_QUESTIONS_LIST.length }
		/>

		...
	</Fragment>
);
```

Also, please, make sure the progress bar has the same styles as it had when it was in the preview component.

File: assets/js/components/user-input/UserInputQuestionnaire.js
Comment (eugene-manuilov) [inline assets/js/components/user-input/UserInputQuestionnaire.js:130]: Let's move this `useQueryArg` hook to other query arg hooks defined above:

```javascript
	const [ activeSlug, setActiveSlug ] = useQueryArg( 'question', steps[ 0 ] );
	const [ redirectURL ] = useQueryArg( 'redirect_url' );
	const [ single ] = useQueryArg( 'single', false );
```

**Review by eugene-manuilov — CHANGES_REQUESTED** [review]
Thanks, @benbowler! It's better now but there are a few more things to address.

File: assets/js/components/user-input/UserInputPreview.js
Comment (benbowler) [reply assets/js/components/user-input/UserInputPreview.js:63]: Are you happy with the error message displaying here outside the questions box? If we want it inside the box we will need to pass it as a prop to `UserInputQuestionWrapper` instead.

Screenshot 2021-01-05 at 16 59 23

**Comment by benbowler** [comment]
Hey @eugene-manuilov, I've updated everything else and the `ErrorNotice` component looks like so:

Screenshot 2021-01-05 at 17 14 56

File: assets/js/components/user-input/UserInputPreview.js
Comment (benbowler) [reply assets/js/components/user-input/UserInputPreview.js]: I've inlined this now so that it looks like the following:

Screenshot 2021-01-06 at 10 38 32
Screenshot 2021-01-06 at 10 39 21

File: assets/js/components/user-input/UserInputPreview.js
Comment (benbowler) [reply assets/js/components/user-input/UserInputPreview.js]: I wasn't able to do this because the `useQueryArgs` hook doesn't update in the parent component when set in the child component. We'd need to work on the hook some more to get this to work as expected.

**Comment by benbowler** [comment]
Hey @eugene-manuilov, back with you. I inlined the errors into each component here's how it looks:

Screenshot 2021-01-06 at 10 38 32
Screenshot 2021-01-06 at 10 39 21

I wasn't able to complete the change to [set the single URL argument in the child component because the hook doesn't currently work as expected there](https://github.com/google/site-kit-wp/pull/2573#discussion_r552508347).

**Comment by benbowler** [comment]
All the rest of the changes have been made.

**Comment by benbowler** [comment]
I can also confirm that the design of the `ProgressBar` component still matches `develop`:

2499-edit-specific-questions-user-input

**Comment by eugene-manuilov** [comment]
> I can also confirm that the design of the `ProgressBar` component still matches `develop`:
> 
> 2499-edit-specific-questions-user-input

I mean the progress bar which appears when you save settings, not the progress at the top which hasn't been affected by your changes.

In the develop branch:
develop

In your branch:
after

**Comment by benbowler** [comment]
@eugene-manuilov I've also updated the ProgressBar to show as it does in develop on save.

**Comment by benbowler** [comment]
JS test failing is from [base branch](https://github.com/google/site-kit-wp/pull/2476).

File: assets/js/components/user-input/UserInputQuestionnaire.js
Comment (eugene-manuilov) [inline assets/js/components/user-input/UserInputQuestionnaire.js:159]: Let's assign it to a variable and use it to escape duplicates:

```javascript
const settingsProgress = (
    <ProgressBar
        height={ 0 }
        indeterminate={ false }
        progress={ ( activeSlugIndex + 1 ) / USER_INPUT_QUESTIONS_LIST.length }
    />
);

if ( isSavingSettings || isNavigating ) {
    return (
        <Fragment>
            { settingsProgress }
            <div> ... </div>
        </Fragment>
    );
}

return (
    <Fragment>
        { settingsProgress }
        ...
    </Fragment>
);
```

File: assets/js/components/user-input/UserInputQuestionnaire.js
Comment (eugene-manuilov) [inline assets/js/components/user-input/UserInputQuestionnaire.js:275]: This is not needed because if either `isSavingSettings` or `isNavigating` is true, we will return earlier.

```suggestion
			{ activeSlug === 'preview' && (
```

**Review by eugene-manuilov — CHANGES_REQUESTED** [review]
Thanks, @benbowler! The last two comments from me.

**Comment by benbowler** [comment]
@eugene-manuilov all requested changes resolved, thanks for working through this with me.

**Review by eugene-manuilov — APPROVED** [review]
Awesome! Thanks, @benbowler!

**Comment by eugene-manuilov** [comment]
@benbowler i am going to leave this ticket in the Code Review column until the PR for #2041 is merged.

**Review by felixarntz — APPROVED** [review]
Awesome work!

**Pull request merged**
